drivers/winc1500/include/winc1500_internal.h

```diff
@@ -0,0 +1,168 @@
+#define WINC1500_EVENT_MBOX_LEN        (1)
+/** @} */
+
+/**
+ * @name   Flags for device internal states (see datasheet)
+ * @{
+ */
+#define WINC1500_STATE_INIT          (1 << 0) /**< Driver is initialized */
+#define WINC1500_STATE_IDLE          (1 << 1) /**< idle state after init */
+#define WINC1500_STATE_STA           (1 << 2) /**< connected to an AP */
+#define WINC1500_STATE_CONNECTED     (1 << 3) /**< connected to an AP */
+#define WINC1500_STATE_IP_OBTAINED   (1 << 4) /**< IP is obtained from DHCP server */
+/** Others are not used yet */
+#define WINC1500_STATE_AP            (1 << 5) /**< Reserved. Wi-Fi AP mode */
+#define WINC1500_STATE_SNIFFER       (1 << 6) /**< Reserved. WINC1500 sniffer mode */
+#define WINC1500_STATE_P2P           (1 << 7) /**< Reserved. P2P mode (Wi-Fi Direct) */
+/** @} */
+
+/**
+ * @brief   Flags for device internal states (see datasheet)
+ */
+typedef enum {
+    WINC1500_EVENT_NOTHING =                   (0),
+    WINC1500_EVENT_OTHERS =                    (1<<0),
+    /* From M2M_WIFI_RESP_CON_STATE_CHANGED */
+    WINC1500_EVENT_CON_STATE_CONNECTED =       (1<<1),
+    WINC1500_EVENT_CON_STATE_DISCONNECTED =    (1<<2),
+    /* From M2M_WIFI_RESP_SCAN_DONE */
+    WINC1500_EVENT_SCAN_DONE =                 (1<<3),
+    /* From M2M_WIFI_RESP_SCAN_RESULT */
+    WINC1500_EVENT_SCAN_RESULT =               (1<<4),
+    /* From M2M_WIFI_RESP_CURRENT_RSSI */
+    WINC1500_EVENT_CURRENT_RSSI =              (1<<5),
+    /* From M2M_WIFI_RESP_CONN_INFO */
+    WINC1500_EVENT_CONN_INFO =                 (1<<6)
+} winc1500_cb_msg_t;
+
+/**
+ * @brief   Flags for device internal states (see WINC1500 programming guide)
+ *          Each types has a corresponding event.
+ */
+typedef union {
+    tstrM2mWifiStateChanged state_change;   /**< M2M_WIFI_RESP_CON_STATE_CHANGED */
+    tstrSystemTime          sys_time;       /**< M2M_WIFI_RESP_GET_SYS_TIME */
+    tstrM2MConnInfo         conn_info;      /**< M2M_WIFI_RESP_CONN_INFO */
+    tstrM2MIPConfig         ip_config;      /**< M2M_WIFI_REQ_DHCP_CONF */
+    tstrM2MWPSInfo          wps_info;       /**< M2M_WIFI_REQ_WPS */
+    uint32_t                ip_conflicted;  /**< M2M_WIFI_RESP_IP_CONFLICT */
+    tstrM2mScanDone         scan_done;      /**< M2M_WIFI_RESP_SCAN_DONE */
+    tstrM2mWifiscanResult   scan_result;    /**< M2M_WIFI_RESP_SCAN_RESULT */
+    tstrM2MProvisionInfo    prov_info;      /**< M2M_WIFI_RESP_PROVISION_INFO */
+    tstrM2MDefaultConnResp  default_conn_resp;  /**< M2M_WIFI_RESP_DEFAULT_CONNECT */
+    tstrPrng                prng_result;        /**< M2M_WIFI_RESP_GET_PRNG */
+    uint8_t                 rx_buf[8];      /**< M2M_WIFI_RESP_CURRENT_RSSI and
+                                                 M2M_WIFI_RESP_CLIENT_INFO */
+#ifdef MODULE_GNRC_NETDEV
+    tstrM2mIpRsvdPkt        recv_pkt;   /**< M2M_WIFI_RESP_ETHERNET_RX_PACKET */
+#endif
+} winc1500_event_info_t;
+
+extern winc1500_t winc1500; /**< winc1500 object */
```

Comment:
I think this is a problem since it prevents having multiple devices. This is very unlikely to happen but is not the way other device drivers are handled by RIOT.
In the auto_init file, there should a static array of device descriptor defined. Only use devices descriptor from there.

Reply:
Yes I don't like this either...please see my follow-up comment in the discussion. Should I force making it as array regardless of the Atmel's WINC1500 API limitation?